### PR TITLE
ThreadPool clean up : mm_pause in loops, correctly spin-then-wait, and adopt static methods consistently in the API

### PR DESCRIPTION
**Pull request by tlh20**
**Description**:  This change makes three changes to the ThreadPool class to clean up issues identified during performance analysis and optimization.  (1) It uses mm_pause intrinsics in spin loops, helping avoid consuming pipeline resources while waiting.  (2) It re-organizes the spin-then-steal loop for work distribution to start out spinning as intended, rather than to start out trying to steal.  (3) It updates the ThreadPool class's API to be consistent in the use of static methods for public functions.  The PR includes minor doc updates and corresponding changes to test cases.

**Motivation and Context**
The change helps ensure consistency in behavior between the OpenMP and Eigen-based implementations.  Unlike the instance methods, the static methods abstract over the different ways in which threading can be implemented; they will map onto the OpenMP or Eigen-based implementations when threading is used.  When threading is not used they will run work sequentially.

### CODE REVIEW

File: include/onnxruntime/core/common/spin_pause.h [hunk 1]
```diff
@@ -0,0 +1,24 @@
+// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License.
+
+#pragma once
+
+#if defined(__x86_64__)
```
Comment (tracysh) [inline include/onnxruntime/core/common/spin_pause.h:6]: This should be the below to make it work with MSVC too:

#if defined(_M_AMD64) || defined(__x86_64__)

Comment (tlh20) [reply include/onnxruntime/core/common/spin_pause.h:6]: Thanks; I has a related question I was going to check on before submitting -- do you know an equivalent to mm_pause we should use on ARM?

Comment (tracysh) [reply include/onnxruntime/core/common/spin_pause.h:6]: ARM does have a yield instruction that is a nop hint to the processor that the current hardware thread can be swapped out. MSVC has a __yield intrinisic to use this. With gcc/clang, you can do __asm__ __volatile__ ("yield");. I don't have an ARM64 system that has multiple hardware threads to measure the impact of this. Maybe just leave it out for now?

File: include/onnxruntime/core/common/spin_pause.h [hunk 2]
```diff
@@ -0,0 +1,24 @@
...
+#endif
+
+namespace onnxruntime {
+
+namespace concurrency {
+
+// Intrinsic to use in spin-loops
+
+inline void SpinPause() {
+#if defined(__x86_64__)
```
Comment (tracysh) [inline include/onnxruntime/core/common/spin_pause.h:17]: This needed to be changed too for _M_AMD64.

**Review by snnn — APPROVED** [review]

**Pull request merged**